### DISCUSSION TITLE
option功能怎么通入编译输入参数添加指定文件夹的xmake脚本

### DISCUSSION THREAD

**Discussion by lucas-wu-wakakak** [xmake-io/xmake#2980] [Q&A] [answered]
有多个项目共用一套代码，现在需要根据xmake --var=dir 来添加指定dir下的子脚本进行分项目编译。
现在有试过通过set_config("cmp_dir","var")，然后通过$(cmp_dir)来拿取值，目前在target部分命令下是生效的，
但是在includes("$(cmp_dir)/product_xmake.lua")中却无法生效，有其他办法可以实现该目的吗？

**Comment by waruqi** [comment]
干嘛要这么折腾，`includes("**/xmake.lua")` 一次性引入所有，子项目 target 配置 set_default(false) 默认不编译。

然后 `xmake build xxx` 想编译哪个，就编译哪个 target 

也可以将 target 通过 set_group 分组设置，按 `xmake -g test_*` 根据组名，模式匹配，批量构建。。随便怎么搞。。

**Reply by lucas-wu-wakakak** [reply]
我担心target太多有点影响编译效率，每次编译都需要轮训遍历。
还有一个问题是如果走多个target模式，各个target的路径都很深，然后包含公用头文件的话，很多文件相对路径会非常长，现在实测好像会影响编译速度，大概15%的差距，所以想共用一个target，临时包含少部分路径特别长的项目文件。

**Reply by lucas-wu-wakakak** [reply]
如果是多target模式，我其实可以用子target在上层，包含公共xmake.lua的形式，现在主要就是路径依赖很难受，大量的公共仓库文件相对路口特别长。

**Reply by waruqi** [reply]
> 我担心target太多有点影响编译效率，每次编译都需要轮训遍历。

理论上不会，只有实际参与编译用到的 target 才会加载，即使你指定 include 特定目录，改需要编译的那些 targets ，还是那些，又不会变多变少。。

**Reply by waruqi** [reply]
> 如果是多target模式，我其实可以用子target在上层，包含公共xmake.lua的形式，现在主要就是路径依赖很难受，大量的公共仓库文件相对路口特别长。

每个子项目一个 xmake.lua，里面只配置自己子项目的相对文件路径。。includes("**/xmake.lua") 批量一次引入，又不需要挨个指定全路径。。

**Reply by lucas-wu-wakakak** [reply]
子xmake.lua里面目前用的是函数形式在函数里去包含各自所管辖的c/h文件，里面的路径目前实测是基于target_run_dir的相对路径, 借用子xmake.lua的当前路径target编译时会漏编译源文件而报错

**Reply by waruqi** [reply]
> 借用子xmake.lua的当前路径target编译时会漏编译源文件而报错

肯定不会，除非你没配置对，推荐做法就是基于当前 xmake.lua 路径配置，你这种搞法就是把简单配置弄复杂化了。。

**Reply by lucas-wu-wakakak** [reply]
用函数的形式是因为假如在子项目里面全局包含的话，那么通过正则方式全部包含，那么各个项目直接所有的源文件都会包进来，然后编译会冲突的。

**Reply by waruqi** [reply]
> 用函数的形式是因为假如在子项目里面全局包含的话，那么通过正则方式全部包含，那么各个项目直接所有的源文件都会包进来，然后编译会冲突的。

没懂，add_files 指定配置自己需要的文件不就好了，跟是否函数里面搞没啥区别，而且 add_files 在模式匹配的同时，可以忽略一批文件，也可以 remove_files 对特定 target 剔除一部分不需要的文件。

另外，干嘛要全局配置，每个 target 单独配置不就好了，通用的才放全局，跟是否函数 有啥关系

**Reply by lucas-wu-wakakak** [reply]
> > 借用子xmake.lua的当前路径target编译时会漏编译源文件而报错
> 
> 肯定不会，除非你没配置对，推荐做法就是基于当前 xmake.lua 路径配置，你这种搞法就是把简单配置弄复杂化了。。

在子xmake.lua里采用函数形式，定义函数去包含，然后在顶层xmake.lua去调用函数，子函数的路径生效的真的是target_rundir的路径，开始我以为也是子xmake.lua的路口，后面debug才发现是调用上层出的路径。

**Reply by lucas-wu-wakakak** [reply]
> 指定配置自己需要的文件不就好了，跟是否函数里面搞没啥区别，而且 add_files 在模式匹配的同时，可以忽略一批文件，也可以 remove_files 对特定 target 剔除一部分不需要

如果单独去remove 那么就需要知道当前所有其他项目的情况，根据其他项目的参数一个一个去remove，那样弄起来也很麻烦的。

**Reply by lucas-wu-wakakak** [reply]
> 另外，干嘛要全局配置，每个 target 单独配置不就好了，通用的才放全局，跟是否函数 有啥关系

子项目会有各自的项目配置文件去配置共用公用中哪些功能会开启需要编译，所以目前是子项目会有一个大的配置文件，里面用set_config去配置大量宏，然后公用仓库的各个模块会去用is_config检查自己管辖的config是否生效，is_config目前在公用仓库模块的lua脚本中只能在函数中才能正常生效，这也是公用模块采用函数管理而不是采用全局成效形式的一个主要原因

**Reply by waruqi** [reply]
> > 指定配置自己需要的文件不就好了，跟是否函数里面搞没啥区别，而且 add_files 在模式匹配的同时，可以忽略一批文件，也可以 remove_files 对特定 target 剔除一部分不需要
> 
> 如果单独去remove 那么就需要知道当前所有其他项目的情况，根据其他项目的参数一个一个去remove，那样弄起来也很麻烦的。

还是没懂你要搞啥，自己子项目目录组织好，每个子项目一个单独的目录维护，按理 add_files("**.cc") 都足够了，压根用不到 remove_files ，除非你每个子项目还混了乱七八糟其他子项目源码。。即使这样，基于当前子目录 `add_files("../xxx/xx.cc")` 单独子目录引入 也就好了。

**Reply by lucas-wu-wakakak** [reply]
> > > 指定配置自己需要的文件不就好了，跟是否函数里面搞没啥区别，而且 add_files 在模式匹配的同时，可以忽略一批文件，也可以 remove_files 对特定 target 剔除一部分不需要
> > 
> > 
> > 如果单独去remove 那么就需要知道当前所有其他项目的情况，根据其他项目的参数一个一个去remove，那样弄起来也很麻烦的。
> 
> 还是没懂你要搞啥，自己子项目目录组织好，每个子项目一个单独的目录维护，按理 add_files("**.cc") 都足够了，压根用不到 remove_files ，除非你每个子项目还混了乱七八糟其他子项目源码。。即使这样，基于当前子目录 `add_files("../xxx/xx.cc")` 单独子目录引入 也就好了。

大哥大哥，项目特殊需要，这样弄真的是为了考虑解耦和兼容后续新项目考虑的，要保证后续新建项目xmake脚本一行代码也不改动就能兼容。总之就是我很清楚这么做的出发点。
有没有我标题说的那种功能[流泪笑]

**Reply by waruqi** [reply]
option + if has_config() then includes()

**Reply by lucas-wu-wakakak** [reply]
> option + if has_config() then includes()

option 怎么样转换成宏或者字符呢？能够在string中生效，类似于var.$(curdir)这种[捂脸笑]

**Reply by waruqi** [reply]
$(curdir) 其他 add_xxx 接口可以直接用，但是 includes 这个里面用不会生效

**Reply by lucas-wu-wakakak** [reply]
ok,感谢~

**Comment by lucas-wu-wakakak** [comment, marked as answer]
我可能偏离重点了，有没有编译传入的option参数变成环境宏的方法，这个环境宏可以用来配置路径，在string中能够替换生效？
--
目前暂时没有更完美解决此问题的方法
